mqtt/mqtt-bridge/src/persist/waking_state/ring_buffer/mod.rs

```diff
@@ -511,30 +511,44 @@ where
     T: Read + Seek,
 {
     // We don't need any explicit returns elsewhere as EOF should be an ERR.
+
+    // There shouldn't be any issues from cast as block size is less than u64 max and usize max always.
     #[allow(clippy::cast_possible_truncation)]
-    let all_zeroes: Vec<u8> = vec![0; *SERIALIZED_BLOCK_SIZE as usize];
-    loop {
-        #[allow(clippy::cast_possible_truncation)]
-        let mut buf = vec![0; *SERIALIZED_BLOCK_SIZE as usize];
-        readable.read_exact(&mut buf)?;
+    let serialized_block_size = *SERIALIZED_BLOCK_SIZE as usize;
+    let buffer_size = (serialized_block_size * 2) as usize;
 
-        // If all zeroes we can skip faster.
+    // Read two blocks worth of data and scan to see if there is a block.
+    // If not, then read another block worth of data and continue...
+    let mut buf = vec![0; buffer_size];
+    readable.read_exact(&mut buf)?;
+
+    let all_zeroes = vec![0; buffer_size];
+    loop {
         if buf == all_zeroes {
-            // This is okay since the block size is too small to wrap.
-            #[allow(clippy::cast_possible_wrap)]
-            readable.seek(SeekFrom::Current(*SERIALIZED_BLOCK_SIZE as i64))?;
+            readable.read_exact(&mut buf)?;
             continue;
         }
 
-        if let Ok(block) = bincode::deserialize::<BlockHeaderWithCrc>(&buf) {
-            let BlockVersion::Version1(inner) = block.inner();
-            // We maybe found a block.
-            if inner.hint() == BLOCK_HINT {
-                return Ok(block);
+        // Try to find block in the buffer we have, if cannot find it by halfway point,
+        // we will need to fetch more data.
+        for offset in 0..(serialized_block_size) {
```

Comment:
just to confirm, our blocks are not aligned, so you need to check every single byte (and not 0,4,8,12,16, for example)?

Reply:
Yeah, I can double check, but I think block header size is an odd number (41 or 43?) thus cannot do offsets of 4,8, etc...